packages/bsky/src/auth.ts

```diff
@@ -30,6 +30,45 @@ export const authOptionalVerifier =
     return authVerifier(idResolver, opts)(reqCtx)
   }
 
+export const authOptionalAccessOrRoleVerifier = (
+  idResolver: IdResolver,
+  cfg: ServerConfig,
+) => {
+  const verifyAccess = authVerifier(idResolver, { aud: cfg.serverDid })
+  const verifyRole = roleVerifier(cfg)
+  return async (ctx: { req: express.Request; res: express.Response }) => {
+    const defaultUnAuthorizedCredentials = { credentials: { did: null } }
+    if (!ctx.req.headers.authorization) {
+      return defaultUnAuthorizedCredentials
+    }
+    // For non-admin tokens, we don't want to consider alternative verifiers and let it fail if it fails
+    const isRoleAuthToken = ctx.req.headers.authorization?.startsWith(BASIC)
+    if (isRoleAuthToken) {
+      const result = await verifyRole(ctx)
+      return {
+        ...result,
+        credentials: {
+          type: 'role' as const,
+          ...result.credentials,
+        },
+      }
+    }
+    try {
+      const result = await verifyAccess(ctx)
+      return {
+        ...result,
+        credentials: {
+          type: 'access' as const,
+          ...result.credentials,
+        },
+      }
+      // If access check fails, we can just return unauthorized credentials since this verifier is optional
+    } catch (err) {
+      return defaultUnAuthorizedCredentials
+    }
```

Comment:
If it's no big deal, I lean towards allowing this to fail if auth is present but invalid.